planner/core/find_best_task.go

```diff
@@ -849,7 +863,110 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 	return task, nil
 }
 
-func (ts *PhysicalTableScan) addPushedDownSelection(copTask *copTask, stats *property.StatsInfo) {
+// pushDownSelAndResolveVirtualCols push some filters down to coprocessor and resolve virtual columns.
+//	1. push down filters and check if there is virtual column.
+//	2. substitute virtual columns in TableScan's Schema to corresponding physical columns.
+//	3. substitute virtual columns in TableScan's filters and push them down.
+//	4. add a Projection upon this DataSource.
+func (ds *DataSource) pushDownSelAndResolveVirtualCols(copTask *copTask, path *accessPath, stats *property.StatsInfo) (t task, err error) {
+	// step 1
+	t = copTask
+	if copTask.indexPlan != nil {
+		if err := ds.addPushedDownSelection(copTask, path); err != nil {
+			return invalidTask, err
+		}
+	}
+	if copTask.tablePlan == nil { // don't need to handle virtual columns in IndexScan
+		return
+	}
+	ts, ok := copTask.tablePlan.(*PhysicalTableScan)
```

Comment:
Just for safety... 🤣 
It's an implicit assumption that the `tablePlan` here must be a `PhysicalTableScan` and it may be broken by some lines of new code added by someone who doesn't know this implicit assumption.

Reply:
By using `ts := copTask.tablePlan.(*PhysicalTableScan)`, if new code breaks the assumption, it would panic then so existing tests can protect us?